kratos/utilities/strategies_utilities.h

```diff
@@ -0,0 +1,230 @@
+
+        // The diagonal considered
+        const double scale_factor = GetScaleNorm<TSparseSpace>(rModelPart, rA, ScalingDiagonal);
+
+        // Detect if there is a line of all zeros and set the diagonal to a 1 if this happens
+        IndexPartition<std::size_t>(system_size).for_each([&](std::size_t Index){
+            bool empty = true;
+
+            const std::size_t col_begin = Arow_indices[Index];
+            const std::size_t col_end = Arow_indices[Index + 1];
+
+            for (std::size_t j = col_begin; j < col_end; ++j) {
+                if(Avalues[j] != 0.0) {
+                    empty = false;
+                    break;
+                }
+            }
+
+            if(empty) {
+                rA(Index, Index) = scale_factor;
+                rb[Index] = 0.0;
+            }
+        });
+
+        return scale_factor;
+    }
+
+    /**
+     * @brief This method returns the scale norm considering for scaling the diagonal
+     * @param rModelPart The problem model part
+     * @param rA The LHS matrix
+     * @param ScalingDiagonal The type of caling diagonal considered
+     * @return The scale norm
+     */
+    template<class TSparseSpace>
+    static double GetScaleNorm(
+        ModelPart& rModelPart,
+        typename TSparseSpace::MatrixType& rA,
+        const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING
+        )
+    {
+        switch (ScalingDiagonal) {
+            case SCALING_DIAGONAL::NO_SCALING:
+                return 1.0;
+            case SCALING_DIAGONAL::CONSIDER_PRESCRIBED_DIAGONAL: {
+                const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+                KRATOS_ERROR_IF_NOT(r_current_process_info.Has(BUILD_SCALE_FACTOR)) << "Scale factor not defined at process info" << std::endl;
+                return r_current_process_info.GetValue(BUILD_SCALE_FACTOR);
+            }
+            case SCALING_DIAGONAL::CONSIDER_NORM_DIAGONAL:
+                return GetDiagonalNorm<TSparseSpace>(rA)/static_cast<double>(rA.size1());
+            case SCALING_DIAGONAL::CONSIDER_MAX_DIAGONAL:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+//                 return TSparseSpace::TwoNorm(rA)/static_cast<double>(rA.size1());
+            default:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+        }
+    }
+
+    /**
+     * @brief This method returns the diagonal norm considering for scaling the diagonal
+     * @param rA The LHS matrix
+     * @return The diagonal norm
+     */
+    template<class TSparseSpace>
+    static double GetDiagonalNorm(typename TSparseSpace::MatrixType& rA)
+    {
+        double* Avalues = rA.value_data().begin();
+        std::size_t* Arow_indices = rA.index1_data().begin();
+        std::size_t* Acol_indices = rA.index2_data().begin();
+
+        const double diagonal_norm = IndexPartition<std::size_t>(TSparseSpace::Size1(rA)).for_each<SumReduction<double>>([&](std::size_t Index){
```

Comment:
i cannot do a full review now, however this sprcific function can be made faster by taking the norm of the values vector without looking into cols and rows.

we might even want to use std parallel functions to do it in parallel since we are now using c++17

Reply:
Right now I am moving to an independent utility the code to be reused, I don't want to do a huge refactor (only the parts I fell trivial), I would refactor that functions properly in a future PR.

Reply:
BTW, currently still using C++11, until no official migration I will not use C++17

Reply:
We are now using C++17, what was your idea @RiccardoRossi ?

Reply:
i thought we were computing the norm of them matrix, smthg for which we do not look into the js

Reply:
What is Js?

Reply:
the columns. But anyhow, i am telling that my comment was wrong

Reply:
Ah Ok, i->row, j->column.